### ISSUE TITLE
YARA Python segmentation fault

### ISSUE THREAD

**Issue by kissgyorgy** [IoT-Inspector/unblob#97]
If you download the latest ubuntu LTS, unblob will crash with segmentation fault.

**Comment by kukovecz** [comment]
Hopefully https://github.com/VirusTotal/yara/pull/1615 will solve this as well.

**Comment by kukovecz** [comment]
You can download the latest ubuntu LTS here: https://ubuntu.com/download/desktop

I just downloaded the `ubuntu-20.04.3-desktop-amd64.iso` which triggered segfault.

**Comment by QKaiser** [comment]
Oooh I thought I needed to execute Yara **on** Ubuntu LTS not **against** 🤓

**Comment by QKaiser** [comment]
I have to investigate a bit more but I managed to find a unique rule that triggers the segfault: the one for squashfs_v4. Interestingly it's the only one so far, all the others even squashfs_v3 do not trigger this.

To reproduce it you can craft the following Yara rule file:

```yara
rule squashfs_v4 {
    strings:
        $squashfs_v4_magic_le = { 68 73 71 73 [24] 04 00 }
    condition:
        $squashfs_v4_magic_le
}
```

And run the following Python file:

```python
import sys
import yara

rules = yara.compile(filepath='sqshv4.yara')
matches = rules.match(sys.argv[1], timeout=600)
print(matches)
```

When you do so, you'll always hit the same error:

```
gdb python3
gef > run yaratest.py /mnt/host/ubuntu-20.04.2.0-desktop-amd64.iso
[ Legend: Modified register | Code | Heap | Stack | String ]
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── registers ────
$rax   : 0xffffffff        
$rbx   : 0x00007ffff638d010  →  0x000000000002124b
$rcx   : 0x000000000090b860  →  0x000000000000004b ("K"?)
$rdx   : 0x161             
$rsp   : 0x00007fffffffcbd0  →  0x00000000009a42e0  →  0x00000000ab7f0000
$rbp   : 0x00007ffff6cd2800  →  0x0000000000000001
$rsi   : 0x000000000090ba10  →  0x0000000000001149
$rdi   : 0x0               
$rip   : 0x00007ffff6b9793d  →  <yara_callback+1149> sub QWORD PTR [r15], 0x1
$r8    : 0x00007ffff6be6000  →  0x0000000000000014
$r9    : 0x4330            
$r10   : 0x0               
$r11   : 0x00007fffffffc735  →  0x0e77ac6100333533 ("353"?)
$r12   : 0x00007ffff6bcf6c3  →  0x00294f2c732c4c28 ("(L,s,O)"?)
$r13   : 0x00007fff4a508020  →  0x0000000000000000
$r14   : 0x0               
$r15   : 0x0               
$eflags: [zero carry PARITY adjust SIGN trap INTERRUPT direction overflow RESUME virtualx86 identification]
$cs: 0x0033 $ss: 0x002b $ds: 0x0000 $es: 0x0000 $fs: 0x0000 $gs: 0x0000 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── stack ────
0x00007fffffffcbd0│+0x0000: 0x00000000009a42e0  →  0x00000000ab7f0000	 ← $rsp
0x00007fffffffcbd8│+0x0008: 0x00007ffff658f010  →  0x0000000100000000
0x00007fffffffcbe0│+0x0010: 0x0000000000000000
0x00007fffffffcbe8│+0x0018: 0x0000000104000000
0x00007fffffffcbf0│+0x0020: 0x00007ffff6d41080  →  0x0000000000000001
0x00007fffffffcbf8│+0x0028: 0x00007ffff6c51140  →  0x0000000000000001
0x00007fffffffcc00│+0x0030: 0x0000000100000003
0x00007fffffffcc08│+0x0038: 0x00007ffff6d495c0  →  0x0000000000000001
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x7ffff6b97932 <yara_callback+1138> mov    rsi, rax
   0x7ffff6b97935 <yara_callback+1141> mov    r14, rax
   0x7ffff6b97938 <yara_callback+1144> call   0x7ffff6b961d0 <PyList_Append@plt>
 → 0x7ffff6b9793d <yara_callback+1149> sub    QWORD PTR [r15], 0x1
   0x7ffff6b97941 <yara_callback+1153> jne    0x7ffff6b978f0 <yara_callback+1072>
   0x7ffff6b97943 <yara_callback+1155> mov    rdi, r15
   0x7ffff6b97946 <yara_callback+1158> call   0x7ffff6b96300 <_Py_Dealloc@plt>
   0x7ffff6b9794b <yara_callback+1163> sub    QWORD PTR [r14], 0x1
   0x7ffff6b9794f <yara_callback+1167> jne    0x7ffff6b978f6 <yara_callback+1078>
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── source:/usr/include/py[...].h+470 ────
    465	                               PyObject *op)
    466	 {
    467	     (void)filename; /* may be unused, shut up -Wunused-parameter */
    468	     (void)lineno; /* may be unused, shut up -Wunused-parameter */
    469	     _Py_DEC_REFTOTAL;
 →  470	     if (--op->ob_refcnt != 0) {
    471	 #ifdef Py_REF_DEBUG
    472	         if (op->ob_refcnt < 0) {
    473	             _Py_NegativeRefcount(filename, lineno, op);
    474	         }
    475	 #endif
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "python3", stopped 0x7ffff6b9793d in yara_callback (), reason: SIGSEGV
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x7ffff6b9793d → yara_callback(context=0x9a42e0, message=0x1, message_data=0x7ffff658f010, user_data=<optimized out>)
[#1] 0x7ffff6b9f77d → yr_scanner_scan_mem_blocks(scanner=0x9a42e0, iterator=0x7fffffffd600)
[#2] 0x7ffff6b9fce4 → yr_scanner_scan_mem(scanner=<optimized out>, buffer=0x7fff4a597000 "ER\b", buffer_size=0xab7f0000)
[#3] 0x7ffff6b9d3a9 → yr_rules_scan_mem(rules=0x9615c0, buffer=0x7fff4a597000 "ER\b", buffer_size=0xab7f0000, flags=0x0, callback=0x7ffff6b974c0 <yara_callback>, user_data=0x7fffffffd730, timeout=0x258)
[#4] 0x7ffff6b9d479 → yr_rules_scan_file(rules=0x9615c0, filename=<optimized out>, flags=0x0, callback=0x7ffff6b974c0 <yara_callback>, user_data=0x7fffffffd730, timeout=0x258)
[#5] 0x7ffff6b99226 → Rules_match(self=0x7ffff6d226c0, args=<optimized out>, keywords=<optimized out>)
[#6] 0x5f5e79 → PyCFunction_Call()
[#7] 0x5f6a46 → _PyObject_MakeTpCall()
[#8] 0x570f1d → _PyEval_EvalFrameDefault()
[#9] 0x5696da → _PyEval_EvalCodeWithName()
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  c
Continuing.

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
gef➤  q
```

I tested it against randomly generated files of size up to 2GB, the segfault does not appear. Something unique in this ISO file is triggering it.

Note that it is specific to the yara-python library ! You can run the rule or the compiled rule with standalone yara and it works.

```bash
$ yara -C sqshv4.yc /mnt/host/ubuntu-20.04.2.0-desktop-amd64.iso
squashfs_v4 /mnt/host/ubuntu-20.04.2.0-desktop-amd64.iso
$ yara sqshv4.yara /mnt/host/ubuntu-20.04.2.0-desktop-amd64.iso 
squashfs_v4 /mnt/host/ubuntu-20.04.2.0-desktop-amd64.iso
```

**Comment by QKaiser** [comment]
Actually solved by https://github.com/VirusTotal/yara/pull/1615.

If we want the latest version:

```
git clone --recursive https://github.com/VirusTotal/yara-python.git
cd yara-python/yara
git checkout master
cd ..
git checkout v4.1.3
git switch -c v4.1.3
python3 setup.py build
sudo python3 setup.py install
```

**Comment by QKaiser** [comment]
@vlaci as the [PR](https://github.com/VirusTotal/yara/pull/1615) owner, can you ask them to release the Python submodule with the fix ? 😺 

Once released I'll close this issue.

**Comment by vlaci** [comment]
https://github.com/VirusTotal/yara-python/issues/196

**Comment by QKaiser** [comment]
Yay, they upgraded the yara submodule in yara-python (see https://github.com/VirusTotal/yara-python/commit/254d226d4dd594d038046e0fbb0b7557a3570c6a) and closed our ticket ! 🥳

**Comment by QKaiser** [comment]
@vlaci we can get rid of the yara patches in `nix/yara` now :)

**Comment by martonilles** [comment]
bump the dependency to the latest upstream that contains our patch and get rid of our "local" patches & test it ;)
